lib/db/lowlevel.go

```diff
@@ -99,208 +104,638 @@ func (db *Lowlevel) Committed() int64 {
+	if err := id.Unmarshal(cur); err != nil {
+		return 0
+	}
 
-type snap struct {
-	*leveldb.Snapshot
-	db *Lowlevel
+	return id
 }
 
-func (s *snap) NewIterator(slice *util.Range, ro *opt.ReadOptions) iterator.Iterator {
-	return s.db.newIterator(func() iterator.Iterator { return s.Snapshot.NewIterator(slice, ro) })
+func (db *Lowlevel) setIndexID(device, folder []byte, id protocol.IndexID) {
+	bs, _ := id.Marshal() // marshalling can't fail
+	if err := db.Put(db.keyer.GenerateIndexIDKey(nil, device, folder), bs); err != nil && !db.IsClosed(err) {
+		panic("storing index ID: " + err.Error())
+	}
 }
 
-// iter implements iterator.Iterator which allows tracking active iterators
-// and aborts if the underlying database is being closed.
-type iter struct {
-	iterator.Iterator
-	db *Lowlevel
+func (db *Lowlevel) dropMtimes(folder []byte) {
+	db.dropPrefix(db.keyer.GenerateMtimesKey(nil, folder))
 }
 
-func (it *iter) Release() {
-	it.db.iterWG.Done()
-	it.Iterator.Release()
+func (db *Lowlevel) dropFolderMeta(folder []byte) {
+	db.dropPrefix(db.keyer.GenerateFolderMetaKey(nil, folder))
 }
 
-func (it *iter) Next() bool {
-	return it.execIfNotClosed(it.Iterator.Next)
+func (db *Lowlevel) dropPrefix(prefix []byte) {
+	t, err := db.newReadWriteTransaction()
+	if err != nil {
+		l.Debugln("newReadWriteTransaction:", err)
+		return
+	}
+	defer t.close()
+
+	t.deleteKeyPrefix(prefix)
 }
-func (it *iter) Prev() bool {
-	return it.execIfNotClosed(it.Iterator.Prev)
+
+func unmarshalTrunc(bs []byte, truncate bool) (FileIntf, error) {
+	if truncate {
+		var tf FileInfoTruncated
+		err := tf.Unmarshal(bs)
+		return tf, err
+	}
+
+	var tf protocol.FileInfo
+	err := tf.Unmarshal(bs)
+	return tf, err
 }
-func (it *iter) First() bool {
-	return it.execIfNotClosed(it.Iterator.First)
+
+func unmarshalVersionList(data []byte) (VersionList, bool) {
+	var vl VersionList
+	if err := vl.Unmarshal(data); err != nil {
+		l.Debugln("unmarshal error:", err)
+		return VersionList{}, false
+	}
+	if len(vl.Versions) == 0 {
+		l.Debugln("empty version list")
+		return VersionList{}, false
+	}
+	return vl, true
 }
-func (it *iter) Last() bool {
-	return it.execIfNotClosed(it.Iterator.Last)
+
+type errorSuggestion struct {
+	inner      error
+	suggestion string
 }
-func (it *iter) Seek(key []byte) bool {
-	return it.execIfNotClosed(func() bool {
-		return it.Iterator.Seek(key)
-	})
+
+func (e errorSuggestion) Error() string {
+	return fmt.Sprintf("%s (%s)", e.inner.Error(), e.suggestion)
 }
 
-func (it *iter) execIfNotClosed(fn func() bool) bool {
-	it.db.closeMut.RLock()
-	defer it.db.closeMut.RUnlock()
-	if it.db.closed {
-		return false
-	}
-	return fn()
+// unchanged checks if two files are the same and thus don't need to be updated.
+// Local flags or the invalid bit might change without the version
+// being bumped. The IsInvalid() method handles both.
+func unchanged(nf, ef FileIntf) bool {
+	return ef.FileVersion().Equal(nf.FileVersion()) && ef.IsInvalid() == nf.IsInvalid()
```

Comment:
The diff of this file is unreadable.
Idk, perhaps make it over several commits or something to make it possible to understand what's yours and what's just moving.

The whole "assume no error" looks like a disasster to me.